// @appland/navie/dist/commands/context-command.js
"use strict";
var __await = (this && this.__await) || function (v) { return this instanceof __await ? (this.v = v, this) : new __await(v); }
var __asyncGenerator = (this && this.__asyncGenerator) || function (thisArg, _arguments, generator) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var g = generator.apply(thisArg, _arguments || []), i, q = [];
    return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
    function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
    function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
    function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
    function fulfill(value) { resume("next", value); }
    function reject(value) { resume("throw", value); }
    function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const js_yaml_1 = require("js-yaml");
const transform_search_terms_1 = __importDefault(require("../lib/transform-search-terms"));
class ContextCommand {
    constructor(options, vectorTermsService, lookupContextService) {
        this.options = options;
        this.vectorTermsService = vectorTermsService;
        this.lookupContextService = lookupContextService;
    }
    execute(request, chatHistory) {
        return __asyncGenerator(this, arguments, function* execute_1() {
            const { question, codeSelection } = request;
            const fence = request.userOptions.isEnabled('fence', true);
            const format = request.userOptions.stringValue('format') || 'yaml';
            const tokenLimit = request.userOptions.numberValue('tokenlimit') || this.options.tokenLimit;
            const transformTerms = request.userOptions.isEnabled('terms', true);
            const aggregateQuestion = [
                ...(chatHistory || [])
                    .filter((message) => message.role === 'user')
                    .map((message) => message.content),
                question,
                codeSelection,
            ]
                .filter(Boolean)
                .join('\n\n');
            const searchTerms = yield __await((0, transform_search_terms_1.default)(transformTerms, aggregateQuestion, this.vectorTermsService));
            const filters = {};
            request.userOptions.populateContextFilters(filters);
            const context = yield __await(this.lookupContextService.lookupContext(searchTerms, tokenLimit, filters));
            let contextStr;
            if (format === 'yaml') {
                contextStr = (0, js_yaml_1.dump)(context);
            }
            else {
                contextStr = JSON.stringify(context, null, 2);
            }
            if (fence)
                yield yield __await(`\`\`\`${format}\n`);
            yield yield __await(contextStr);
            if (fence)
                yield yield __await('```\n');
        });
    }
}
exports.default = ContextCommand;
